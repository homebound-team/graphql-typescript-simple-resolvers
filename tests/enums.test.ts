import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("Enum Types", () => {
  const schemaWithEnums = `
    enum Status {
      ACTIVE
      INACTIVE
      PENDING
      ARCHIVED
    }

    enum Priority {
      LOW
      MEDIUM
      HIGH
      URGENT
    }

    enum UserRole {
      GUEST
      USER
      MODERATOR
      ADMIN
    }

    type User {
      id: ID!
      name: String!
      status: Status!
      role: UserRole!
      priority: Priority
    }

    type Query {
      users(status: Status, role: UserRole): [User!]!
      usersByStatuses(statuses: [Status!]!): [User!]!
    }

    input UserInput {
      name: String!
      status: Status
      role: UserRole!
      priority: Priority
    }
  `;

  it("generates enum types", async () => {
    const schema = createTestSchema(schemaWithEnums);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export enum Status {
  Active = "ACTIVE",
  Inactive = "INACTIVE",
  Pending = "PENDING",
  Archived = "ARCHIVED",
}`);

    expect(code).toContain(`export enum Priority {
  Low = "LOW",
  Medium = "MEDIUM",
  High = "HIGH",
  Urgent = "URGENT",
}`);

    expect(code).toContain(`export enum UserRole {
  Guest = "GUEST",
  User = "USER",
  Moderator = "MODERATOR",
  Admin = "ADMIN",
}`);
  });

  it("uses enums in type definitions", async () => {
    const schema = createTestSchema(schemaWithEnums);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface User {
  id: string;
  name: string;
  status: Status;
  role: UserRole;
  priority: Priority | null | undefined;
}`);
  });

  it("uses enums in input types", async () => {
    const schema = createTestSchema(schemaWithEnums);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface UserInput {
  name: string;
  status?: Status | null | undefined;
  role: UserRole;
  priority?: Priority | null | undefined;
}`);
  });

  it("uses enums in query arguments", async () => {
    const schema = createTestSchema(schemaWithEnums);
    const code = await runPlugin(schema, {
      scalars: {},
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`users: Resolver<{}, QueryUsersArgs, readonly User[]>;`);
    expect(code).toContain(`usersByStatuses: Resolver<{}, QueryUsersByStatusesArgs, readonly User[]>;`);
  });

  it("handles enum value mapping to custom types", async () => {
    const schema = createTestSchema(schemaWithEnums);
    const code = await runPlugin(schema, {
      enumValues: {
        Status: "./enums#StatusEnum",
        Priority: "./enums#PriorityEnum",
      },
      scalars: {},
      mappers: {},
    });

    expect(code).toContain(`import { PriorityEnum, StatusEnum } from "./enums";`);
    // Export syntax may use separate export statements
    expect(code).toContain('export { StatusEnum } from "./enums";');
    expect(code).toContain('export { PriorityEnum } from "./enums";');
  });

  it("snapshots enum generation", async () => {
    const schema = createTestSchema(schemaWithEnums);
    const code = await runPlugin(schema, {
      enumValues: {
        Status: "./enums#StatusEnum",
        Priority: "./enums#PriorityEnum",
      },
      scalars: {},
      mappers: {},
    });
    expect(code).toMatchSnapshot();
  });
});
