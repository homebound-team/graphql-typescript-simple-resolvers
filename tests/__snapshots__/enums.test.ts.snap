// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Enum Types snapshots enum generation 1`] = `
"import { GraphQLResolveInfo } from "graphql";
import { Context } from "./context";
import { PriorityEnum, StatusEnum } from "./enums";

export interface Resolvers {
  Query: QueryResolvers;
  User?: UserResolvers;
}

export type UnionResolvers = {};

export interface QueryResolvers {
  users: Resolver<{}, QueryUsersArgs, readonly User[]>;
  usersByStatuses: Resolver<{}, QueryUsersByStatusesArgs, readonly User[]>;
}

export interface UserResolvers {
  id: Resolver<User, {}, string>;
  name: Resolver<User, {}, string>;
  status: Resolver<User, {}, StatusEnum>;
  role: Resolver<User, {}, UserRole>;
  priority: Resolver<User, {}, PriorityEnum | null | undefined>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface QueryUsersArgs {
  status?: StatusEnum | null | undefined;
  role?: UserRole | null | undefined;
}
export interface QueryUsersByStatusesArgs {
  statuses: readonly StatusEnum[];
}
export interface User {
  id: string;
  name: string;
  status: StatusEnum;
  role: UserRole;
  priority: PriorityEnum | null | undefined;
}

export interface UserInput {
  name: string;
  status?: StatusEnum | null | undefined;
  role: UserRole;
  priority?: PriorityEnum | null | undefined;
}

export { StatusEnum } from "./enums";

export { PriorityEnum } from "./enums";

export enum UserRole {
  Guest = "GUEST",
  User = "USER",
  Moderator = "MODERATOR",
  Admin = "ADMIN",
}
"
`;
